src/System.ServiceModel.Primitives/tests/ServiceModel/DuplexChannelFactoryTest.cs

```diff
@@ -0,0 +1,384 @@
+        Assert.Throws<ArgumentNullException>(() =>
+        {
+            DuplexChannelFactory<IWcfDuplexService> factory = new DuplexChannelFactory<IWcfDuplexService>(callback, binding, endpoint);
+            factory.CreateChannel();
+        });
+    }
+
+    [Fact]
+    public static void CreateChannel_Using_NetTcp_NoSecurity()
+    {
+        WcfDuplexServiceCallback callback = new WcfDuplexServiceCallback();
+        Binding binding = new NetTcpBinding(SecurityMode.None);
+        EndpointAddress endpoint = new EndpointAddress("net.tcp://not-an-endpoint");
+        DuplexChannelFactory<IWcfDuplexService> factory = new DuplexChannelFactory<IWcfDuplexService>(callback, binding, endpoint);
+        factory.CreateChannel(); 
+    }
+
+    [Fact]
+    public static void CreateChannel_Using_Http_NoSecurity()
+    {
+        WcfDuplexServiceCallback callback = new WcfDuplexServiceCallback();
+        Binding binding = new NetHttpBinding(BasicHttpSecurityMode.None);
+        EndpointAddress endpoint = new EndpointAddress(Endpoints.HttpBaseAddress_NetHttp);
+        DuplexChannelFactory<IWcfDuplexService> factory = new DuplexChannelFactory<IWcfDuplexService>(callback, binding, endpoint);
+
+        // Can't cast to IDuplexSessionChannel to IRequestChannel on http
+        var exception = Assert.Throws<InvalidCastException>(() =>
+        {
+            factory.CreateChannel();
+        });
+
+        Assert.Contains("IRequestChannel", exception.Message, StringComparison.OrdinalIgnoreCase);
+    }
+    
+    [Fact]
+    public static void CreateChannel_Of_IDuplexChannel_Using_BasicHttpBinding()
+    {
+        ChannelFactory<IWcfDuplexService> factory = null;
+        ChannelFactory<IWcfDuplexService> factory2 = null;
+        IWcfDuplexService channel = null;
+        IWcfDuplexService channel2 = null;
+
+        WcfDuplexServiceCallback callback = new WcfDuplexServiceCallback();
+
+        try
+        {
+            BasicHttpBinding binding = new BasicHttpBinding();
+            EndpointAddress endpointAddress = new EndpointAddress(BaseAddress.HttpBaseAddress);
+
+            // Create the channel factory for the duplex message exchange pattern.
+            factory = new DuplexChannelFactory<IWcfDuplexService>(callback, binding, endpointAddress);
+            factory2 = new DuplexChannelFactory<IWcfDuplexService>(callback, binding, endpointAddress);
+
+            Assert.Equal(CommunicationState.Created, factory.State);
+            Assert.Equal(CommunicationState.Created, factory2.State);
+
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+            {
+                // Create the channel.
+                channel = factory.CreateChannel();
+                Assert.True(typeof(IWcfDuplexService).GetTypeInfo().IsAssignableFrom(channel.GetType().GetTypeInfo()),
+                    String.Format("Channel type '{0}' was not assignable to '{1}'", channel.GetType(), typeof(IDuplexChannel)));
```

Comment:
I  don't think the IsAssignable test belongs here.  By definition the CreateChannel must return a type assignable to IWcfDuplexService.  It would be a runtime error if the DispatchProxy layer attempted to return something that wasn't.